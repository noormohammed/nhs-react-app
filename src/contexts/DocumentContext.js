import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useRef,
} from "react";

/**
 * DocumentContext for Document details
 */

const DocumentContext = createContext({});

/**
 * Document Provider using the DocumentContext
 * @param {*} children any props/children components passed to the DocumentContextProvider
 * @returns memoized object of document info/details state, set state function
 */
const DocumentContextProvider = ({ children }) => {
  // Once the document info is memoized, we store the info in the localStorage
  // for ease of access and its also economical (can avoid too many backend calls).
  const [documentInfo, setDocumentInfo] = useState(
    window.localStorage.getItem("document_info")
      ? JSON.parse(window.localStorage.getItem("document_info"))
      : null
  );

  const isMountedRef = useRef(null);

  useEffect(() => {
    isMountedRef.current = true;
    const storeDocumentInfo = async () => {
      if (isMountedRef.current) {
        window.localStorage.setItem("document_info", JSON.stringify(documentInfo));
      }
    };
    storeDocumentInfo();

    return () => {
      isMountedRef.current = false;
    };
  }, [documentInfo]);

  return (
    <DocumentContext.Provider value={[documentInfo, setDocumentInfo]}>
      {children}
    </DocumentContext.Provider>
  );
};

const useDocumentContext = () => useContext(DocumentContext);

export { useDocumentContext, DocumentContextProvider };
