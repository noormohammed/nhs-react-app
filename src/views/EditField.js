import React, { useState } from "react";
import { useLocation, useHistory } from "react-router";

import { useDocumentContext } from "contexts/DocumentContext";
import Page from "components/Page";
import Form from "components/form/Form";
import { TextAreaField, SelectField } from "components/form/Fields";

/**
 * EditField Component generates a view to display a form with editable field
 *
 * @component
 * @example
 * return (<EditField />)
 */
const EditField = () => {
  const history = useHistory();
  const location = useLocation();
  const fieldData = location?.state;
  const [documentInfo, setDocumentInfo] = useDocumentContext();
  const [fieldValue, setFieldValue] = useState(
    documentInfo ? documentInfo[fieldData?.name] : ""
  );

  const handleChange = (event) => {
    let targetValue;

    // Any field onChange value gets stored in local state,
    // and pushing it to the context only on form submit.

    if (fieldData?.type === "select" && fieldData?.multiple) {
      // Creating an array of selected options for select-multiple
      targetValue = Array.from(event.target.selectedOptions, (option) =>
        parseInt(option.value, 10)
      );
      if (!targetValue[0]) return;
    } else {
      targetValue = event.target.value;
    }

    setFieldValue(targetValue);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    // Relying on default browser validations for this test app,
    // as writing a validation library is beyond the scope of this test

    if (!fieldValue && fieldData.required) return;

    setDocumentInfo({
      ...documentInfo,
      [fieldData?.name]: fieldValue,
    });

    history?.goBack();
  };

  return (
    <Page title={fieldData?.title || "Enter Value"}>
      <Form onSubmit={handleSubmit} submitButtonName="Submit">
        {fieldData?.type === "textarea" && (
          <TextAreaField onChange={handleChange} value={fieldValue} {...fieldData} />
        )}
        {fieldData?.type === "select" && (
          <SelectField
            onChange={handleChange}
            value={documentInfo[fieldData?.name]}
            {...fieldData}
          />
        )}
      </Form>
    </Page>
  );
};

export default EditField;
