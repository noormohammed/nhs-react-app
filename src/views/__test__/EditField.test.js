import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

import { DocumentContextProvider } from "contexts/DocumentContext";
import EditField from "views/EditField";
import SampleDocumentDetails from "static/sample_document_details";
import userEvent from "@testing-library/user-event";

// Using the modes defined in the environmental variables
const { REACT_APP_DOC_EDIT_FIELD_MODE } = process.env;

const documentInfo = { ...SampleDocumentDetails };
documentInfo["mode"] = REACT_APP_DOC_EDIT_FIELD_MODE;
const setDocumentInfo = jest.fn();

describe("<EditField /> TextArea Tests", () => {
  let container;
  // sample field data for testing, passed into state of history
  const textAreaFieldData = {
    title: "Enter Name",
    type: "textarea",
    required: "required",
    name: "name",
    maxLength: 50,
    rows: 1,
    cols: 50,
  };

  const history = createMemoryHistory();
  history.push("/edit-field", textAreaFieldData);

  beforeEach(() => {
    ({ container } = render(
      <DocumentContextProvider value={[documentInfo, setDocumentInfo]}>
        <Router history={history}>
          <EditField />
        </Router>
      </DocumentContextProvider>
    ));
  });

  afterEach(cleanup);

  test("renders DocumentMenu correctly", () => {
    expect(container).toMatchSnapshot();
  });

  test("renders title correctly", () => {
    expect(screen.getByText(/Enter Name/i)).toBeInTheDocument();
  });

  test("renders a form field", () => {
    expect(screen.findByRole("form")).toBeTruthy();
    expect(container.getElementsByClassName("form")).toHaveLength(1);
  });

  test("renders a textarea field", async () => {
    const textfield = screen.getByTestId(`textarea-field-${textAreaFieldData.name}`);
    expect(screen.findByRole("textarea")).toBeTruthy();
    expect(textfield).toBeInTheDocument();
  });

  test("renders a submit button", () => {
    expect(screen.getByText("Submit")).toBeTruthy();
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test("submit button click on empty textarea value does not yeild", () => {
    const submitBtn = screen.getByText("Submit");
    fireEvent.click(submitBtn);
    expect(history.location.pathname).not.toBe("/document-edit");
    // Browser validation is working.
    // Submit did not complete since the value of the textarea is empty
    // hence the view does not change as tested below.
    expect(history.location.pathname).toBe("/edit-field");
  });

  test("trigger submit button click with valid textarea value", async () => {
    const submitBtn = screen.getByText("Submit");
    const textfield = screen.getByTestId(`textarea-field-${textAreaFieldData.name}`);
    userEvent.type(textfield, "Sample Document");

    fireEvent.click(submitBtn);

    // Submit completes successfully as we pass a valid textarea value
    // hence the view changes and goes back to the previous view.
    // Since we are accessing EditField component directly from root,
    // history,goBack() returns the root view.
    expect(history.location.pathname).toBe("/");
  });
});

describe("<EditField /> Select Tests", () => {
  let container;
  // sample field data for testing, passed into state of history
  const selectFieldData = {
    title: "Select Type",
    type: "select",
    required: "required",
    name: "type",
    options: [
      { id: 1, name: "Type-1" },
      { id: 2, name: "Type-2" },
      { id: 3, name: "Type-3" },
    ],
  };

  const history = createMemoryHistory();
  history.push("/edit-field", selectFieldData);

  beforeEach(() => {
    ({ container } = render(
      <DocumentContextProvider value={[documentInfo, setDocumentInfo]}>
        <Router history={history}>
          <EditField />
        </Router>
      </DocumentContextProvider>
    ));
  });

  afterEach(cleanup);

  test("renders DocumentMenu correctly", () => {
    expect(container).toMatchSnapshot();
  });

  test("renders title correctly", () => {
    expect(screen.getByText(/Select Type/i)).toBeInTheDocument();
  });

  test("renders a form field", () => {
    expect(screen.findByRole("form")).toBeTruthy();
    expect(container.getElementsByClassName("form")).toHaveLength(1);
  });

  test("renders a select field", async () => {
    const textfield = screen.getByTestId(`select-field-${selectFieldData.name}`);
    expect(screen.findByRole("select")).toBeTruthy();
    expect(textfield).toBeInTheDocument();
  });

  test("renders a submit button", () => {
    expect(screen.getByText("Submit")).toBeTruthy();
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test("submit button click on empty select value does not yeild", () => {
    const submitBtn = screen.getByText("Submit");
    fireEvent.click(submitBtn);
    expect(history.location.pathname).not.toBe("/document-edit");
    // Browser validation is working.
    // Submit did not complete since the value of the select is empty
    // hence the view does not change as tested below.
    expect(history.location.pathname).toBe("/edit-field");
  });

  test("trigger submit button click with valid select value", async () => {
    const submitBtn = screen.getByText("Submit");
    const selectfield = screen.getByTestId(`select-field-${selectFieldData.name}`);

    userEvent.selectOptions(selectfield, ["1"]);

    fireEvent.click(submitBtn);

    // Submit completes successfully as we pass a valid selected value
    // hence the view changes and goes back to the previous view.
    // Since we are accessing EditField component directly from root,
    // history,goBack() returns the root view.
    expect(history.location.pathname).toBe("/");
  });
});
