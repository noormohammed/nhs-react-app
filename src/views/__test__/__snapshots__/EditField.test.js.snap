// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditField /> Select Tests renders DocumentMenu correctly 1`] = `
<div>
  <div
    class="main"
  >
    <div
      style="position: fixed; top: 0px; display: flex; flex-direction: row; justify-content: center; align-items: center; width: 100%; height: 50px; background-color: rgb(63, 80, 181); border-bottom: 1px solid #d8d8d8; font-weight: bold; padding: 0px 20px 0px 0px; box-sizing: border-box; color: rgb(255, 255, 255); font-size: 18px;"
    >
      <div
        style="width: 50px;"
      />
      <div
        style="margin-left: auto; margin-right: auto;"
      >
        Select Type
      </div>
      <div
        style="width: 50px;"
      />
    </div>
    <div
      class="content"
    >
      <form
        class="form"
      >
        <fieldset>
          <div
            class="field"
          >
            <select
              data-testid="select-field-type"
              name="type"
              required=""
              title="Select Type"
              type="select"
            >
              <option
                value="1"
              >
                Type-1
              </option>
              <option
                value="2"
              >
                Type-2
              </option>
              <option
                value="3"
              >
                Type-3
              </option>
            </select>
          </div>
        </fieldset>
        <div
          style="text-align: center;"
        >
          <button
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
    <div
      style="display: flex; align-items: center; justify-content: center; width: 100%; height: 40px; background-color: rgb(51, 51, 51); color: rgb(255, 255, 255); position: fixed; bottom: 0px;"
    >
      Footer Menu
    </div>
  </div>
</div>
`;

exports[`<EditField /> TextArea Tests renders DocumentMenu correctly 1`] = `
<div>
  <div
    class="main"
  >
    <div
      style="position: fixed; top: 0px; display: flex; flex-direction: row; justify-content: center; align-items: center; width: 100%; height: 50px; background-color: rgb(63, 80, 181); border-bottom: 1px solid #d8d8d8; font-weight: bold; padding: 0px 20px 0px 0px; box-sizing: border-box; color: rgb(255, 255, 255); font-size: 18px;"
    >
      <div
        style="width: 50px;"
      />
      <div
        style="margin-left: auto; margin-right: auto;"
      >
        Enter Name
      </div>
      <div
        style="width: 50px;"
      />
    </div>
    <div
      class="content"
    >
      <form
        class="form"
      >
        <fieldset>
          <div
            class="field"
          >
            <textarea
              cols="50"
              data-testid="textarea-field-name"
              maxlength="50"
              name="name"
              required=""
              rows="1"
              title="Enter Name"
              type="textarea"
            />
          </div>
        </fieldset>
        <div
          style="text-align: center;"
        >
          <button
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
    <div
      style="display: flex; align-items: center; justify-content: center; width: 100%; height: 40px; background-color: rgb(51, 51, 51); color: rgb(255, 255, 255); position: fixed; bottom: 0px;"
    >
      Footer Menu
    </div>
  </div>
</div>
`;
